### ISSUE TITLE
Improve error handling in Decoder.decodeMapIntoMap()

### ISSUE THREAD

**Issue by thoricha91** [maxmind/MaxMind-DB-Reader-java#126]
Hi team,

As part of an internal project for the company I work for, we are using the MaxMind DB reader without own custom schema. Due to a design oversight, I hit a situation where our MMDB file contained a map of floats but we were expecting a map of integers. When the DB reader attempts to decode the map, a ```ClassCastException``` is thrown when trying to cast float to integer in ```Decoder.decodeMapToMap()```.

```java
java.lang.ClassCastException: Cannot cast java.lang.Double to java.lang.Integer
	at java.base/java.lang.Class.cast(Class.java:3889)
	at com.maxmind.db.Decoder.decodeMapIntoMap(Decoder.java:375)
	at com.maxmind.db.Decoder.decodeMap(Decoder.java:338)
	at com.maxmind.db.Decoder.decodeByType(Decoder.java:162)
	at com.maxmind.db.Decoder.decode(Decoder.java:151)
	at com.maxmind.db.Decoder.decodeMapIntoObject(Decoder.java:429)
	at com.maxmind.db.Decoder.decodeMap(Decoder.java:341)
	at com.maxmind.db.Decoder.decodeByType(Decoder.java:162)
	at com.maxmind.db.Decoder.decode(Decoder.java:151)
	at com.maxmind.db.Decoder.decode(Decoder.java:76)
	at com.maxmind.db.Reader.resolveDataPointer(Reader.java:274)
	at com.maxmind.db.Reader.getRecord(Reader.java:183)
```

In this case, the above ```ClassCastException``` is returned all the way up to the point where ```Reader.getReader()``` was called. 

I notice, however, if I were to create a similar data type discrepancy in a Plain Old Java Object (POJO), the decoder hits a ```DeserializationException``` in ```Decoder.decodeMapIntoObject()```.

```java
Caused by: com.maxmind.db.DeserializationException: Error creating object of type: Location -  argument type mismatch in latitude MMDB Type: java.lang.Double Java Type: java.lang.Integer argument type mismatch in longitude MMDB Type: java.lang.Double Java Type: java.lang.Integer
	at com.maxmind.db.Decoder.decodeMapIntoObject(Decoder.java:452)
	at com.maxmind.db.Decoder.decodeMap(Decoder.java:341)
	at com.maxmind.db.Decoder.decodeByType(Decoder.java:162)
	at com.maxmind.db.Decoder.decode(Decoder.java:151)
	at com.maxmind.db.Decoder.decode(Decoder.java:89)
	at com.maxmind.db.NoCache.get(NoCache.java:17)
	at com.maxmind.db.Decoder.decode(Decoder.java:116)
	at com.maxmind.db.Decoder.decodeMapIntoObject(Decoder.java:429)
	at com.maxmind.db.Decoder.decodeMap(Decoder.java:341)
	at com.maxmind.db.Decoder.decodeByType(Decoder.java:162)
	at com.maxmind.db.Decoder.decode(Decoder.java:151)
	at com.maxmind.db.Decoder.decode(Decoder.java:76)
	at com.maxmind.db.Reader.resolveDataPointer(Reader.java:274)
	at com.maxmind.db.Reader.getRecord(Reader.java:183)
```

In this case, I see the existing error handling nicely converts the above exception into one more meaningful. At the point where ```Reader.getReader()``` was called, I saw:

```java
com.maxmind.db.DeserializationException: Error getting record for IP /<test_ip_address> -  Error creating object of type: Location -  argument type mismatch in latitude MMDB Type: java.lang.Double Java Type: java.lang.Integer argument type mismatch in longitude MMDB Type: java.lang.Double Java Type: java.lang.Integer
	at com.maxmind.db.Reader.getRecord(Reader.java:186)
```

My suggestion would be to wrap the cast function with a try-catch at [Decoder.java#L375](https://github.com/maxmind/MaxMind-DB-Reader-java/blob/470eadb842f8fc57a17bd3ff26bf4a643b2deea5/src/main/java/com/maxmind/db/Decoder.java#L375) to then throw a DeserializationException with a relevant cause and message. This should allow values to be casted where the casting is valid.

Please let me know your thoughts, or if I need to re-clarify anything that is ambiguous. I am additionally happy to contribute any code change as needed.

Thanks!

Thom

**Comment by oschwald** [comment]
Turning the exception into a `DeserializationException` makes sense to me. If you wanted to make a PR, we would be happy to include it.

**Comment by thoricha91** [comment]
Thank you for such a quick reply :-) I'll draw up a PR shortly.

**Comment by thoricha91** [comment]
Here is a first pass PR with my suggested change: https://github.com/maxmind/MaxMind-DB-Reader-java/pull/127

**Comment by oschwald** [comment]
Fixed in #127.
